sdbd.8
.TH SDBD 8 LOCAL

.SH NAME
sdbd

.SH SYNOPSIS
.B sdbd
[-d] [-f] port

.SH DESCRIPTION
.I sdbd
allows applications using libsdb to access databases without linking with
the database client libraries. It also allows access to databases that
are located on a different host, even if they don't have a networking
protocol.

.SH SEE ALSO

Sample client in sdbd_client.c

.SH AUTHOR
Copyright (C) 2000-2002 Ulric Eriksson <ulric@siag.nu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the Licence, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Library General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston,
MA 02111-1307, USA.



ChangeLog

150222	Updated config.in.

101005	More tweaking of the configuration script for compatibility
	with the latest instant oracle client.
	Released 0.10.1.

080430	Oracle driver again: increased the size of the hda to 512 bytes
	for compatibility with 64-bit platforms.
	Released 0.10.0.

080427	Made the Oracle driver build compatible with Oracle Instant Client.

080425	Added db option to oracle url. These are now equivalent:
	oracle:uid=username/password@database
	oracle:uid=username:pwd=password:db=database

080311	Added --disable-shared and --disable-static options to
	the configure script.
	Released 0.9.0.

080307	In sdb.c, function sdb_close: don't call the driver's close
	function if the db is null (i.e. there is nothing to close).

080306	Started integrating Freetds for Microsoft SQL and Sybase.
	Queries are submitted but returned values are not handed
	off to the callback.

080305	Oracle driver updated for 11g.
	Make sure that the Oracle driver treats null values as NULL.
	Released 0.8.0.

080304	Initial support for Ingres.

080229	Added support for shsql/quisp.
	Changed clilen type to socklen_t in sdbd.c.
	Released 0.7.0.

051029	Added -c option to sdb_client to read an sql query from the
	command line rather than interactively.
	Released 0.6.2.

050219	In sdb_query: if the url refers to an already opened connection,
	check that the connection is in fact opened. This is important
	for mistyped urls.
	Released 0.6.1.

050110	Added --platform flag to the configure script.

050110	Added support for Sqlite 3.
	Added another minimal example to the manpage.
	Released 0.6.0.

050109	Dumped automake. Its self-incompatibility drives me nuts!

040506	More bugfixes from Guillaume Cottenceau to the Postgres
	and Mysql drivers: don't segfault if sdb_close is called
	with a NULL argument.
	Released 0.5.5.

040423	Bugfix in the postgres driver: the closer should be called
	sdb_postgres_close. Released 0.5.4.

040421	Patch for persistent Postgres connections from
	Guillaume Cottenceau <gc at mnc.ch>. Released 0.5.3.

040227	Check for Solaris specific libraries.
	Added patch for persistent Oracle connections
	from Denis Ovsienko <pilot@altlinux.ru>.

020828	Do not link sdb_client against .libs/libsdb.so. It breaks the
	build on certain platforms.
	sdb_client: only print prompt if stdin is a terminal.
	sdb_client: Option -s changes the separator in output.
	Released 0.5.1.

020630	Added a driver by Lahcen EL HADDATI <elhaddati@hotmail.com>
	which accepts queries on plain text files in a SQL-like language.
	Released 0.5.0.

020621	Fixed bug in sdbd: null values in the result set caused the
	process to crash. Now it returns an empty string, which is wrong
	but better than crashing.
	Check memory allocations better in sdbd_driver.

020620	Mimer driver that actually works.
	Released 0.4.1.

020530	MiniSQL driver. Use --with-msql=/path/to/msql if it is not
	installed in /usr/local/Hughes.
	Released 0.4.0.

020526	Implemented open and close for mysql driver.
	Wrote manpages for sdb_open, sdb_close, sdb_query, sdb_init and sdbd.
	Released 0.3.0.

020522	Added sdb_open and sdb_close for connections that are to be
	used for more than one query.
	Implemented open and close for sqlite driver.
	Modified the sdbd protocol to allow connections to be used for
	multiple queries. Updated sdbd.c and sdbd_driver.c to use the new
	protocol if it is implemented on both server and client.

020520	Added a check in sdb_query to call sdb_init if it hasn't already
	been done. This makes sdb_init calls from applications optional.

020515	Improved error checking in gdbm_driver.c.

020514	Added squid_sdb_auth.c, a program to authenticate Squid proxy
	users to any database supported by libsdb.
	Released 0.2.4.

020503	Renamed all the drivers from XXX.c to XXX_driver.c.
	Wrote a new driver (sdbd_driver.c) for proxied queries.
	The new driver is designed so it can be compiled statically
	into an application without the rest of libsdb.

020429	Added optional argument to --with-odbc option, plus the necessary
	tweaking to get the odbc driver running on Windows.
	Changed #include <mysql/mysql.h> to just #include <mysql.h>.
	Released 0.2.3.

020425	New flags -d and -f to sdbd control debug message level and
	background/foreground operation.
	Made sdbd less eager to error out at every problem.

020209	Modification to postgres driver to let it work with
	user/password login (srosso@libero.it).
	Released 0.2.2.

020105	Fixed compiler warning in sdbd_client.c (2nd arg to connect).

011220	Updated automake, autoconf and libtool.

011204	Updated configuration for Mysql.

010910	SO_REUSEADDR in sdbd.c.

010611	Fixed silly bug in sdbd.c and sdbd_client.c: readdata and writedata
	would fail when data was read and written in small segments.
	Released 0.2.1.

010610	Sample client for sdbd in sdbd_client. Released 0.2.0.

010601	Proxy daemon called sdbd.

010520	Updated documentation and sdb_client.
	Released 0.1.0.

010517	Added trivial gdbm driver (adapted from Siag).

010515	Added autoconf and libtool.

010513	Released 0.0.1.

010512	Cleaned up namespace by prefixing all globals with sdb_.

010511	Started project by copying database code from Ucronadm.



sdbd.c
/*
   Copyright (C) 2001-2008  Ulric Eriksson <ulric@siag.nu>
 
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the Licence, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include "sdb.h"
#include "common.h"

static int foreground = 0;

static void writedata(int fd, char *b, int n)
{
	int i, j;
	for (i = 0; i < n; i += j) {
		j = write(fd, b+i, n-i);
		if (j < 0) sdb_error("Error writing data");
	}
}

static void writestring(int fd, char *b)
{
	writedata(fd, b, strlen(b));
}

static void waitforchild(int i)
{
	if (sdb_debuglevel) sdb_debug("waitforchild()");
	waitpid(-1, NULL, WNOHANG);
	signal(SIGCHLD, waitforchild);
}

static void chomp(char *p)
{
	int i;
	for (i = 0; p[i]; i++) {
		if (p[i] == '\r' || p[i] == '\n') {
			p[i] = '\0';
			return;
		}
	}
}

static int db_callback(int n, char **p, void *closure)
{
	int i;
	int fd = *((int *)closure);
	char b[10];

	if (!n) return 0;
	sprintf(b, "%d ", n);
	writestring(fd, b);		/* number of columns */
	for (i = 0; i < n; i++) {
		char *q = p[i];
		if (q == NULL) q = "";
		if (sdb_debuglevel > 1) sdb_debug("%s ", q);
		sprintf(b, "%ld ", (long)strlen(q));
		writestring(fd, b);	/* column width */
		writestring(fd, q);	/* column */
	}
	if (sdb_debuglevel > 1) sdb_debug("\n");
	return 0;
}

static int doquery(int sock, char *u)
{
	int n;
	char q[4096];

	if (sdb_debuglevel) sdb_debug("doquery(%d) begin", sock);
	n = read(sock, q, sizeof q);
	if (n < 0) sdb_error("Can't read query");
	q[n] = '\0';
	if (sdb_debuglevel) sdb_debug("'%s'\n", q);

	if (sdb_debuglevel) sdb_debug("calling sdb_query");

	n = sdb_query(u, q, db_callback, &sock);
	writedata(sock, "0", 2);		/* The End */
	if (sdb_debuglevel) sdb_debug("doquery() end");

	return n;
}

static void dourl(int sock)
{
	int n;
	char u[1024], *db;

	if (sdb_debuglevel) sdb_debug("dourl(%d) begin", sock);
	n = read(sock, u, sizeof u);
	if (n < 0) sdb_error("Can't read URL");
	u[n] = '\0';
	chomp(u);
	if (sdb_debuglevel) sdb_debug("'%s'\n", u);
	writedata(sock, " ", 1);

	if (!strcmp(u, "sdb_open")) {
		n = read(sock, u, sizeof u);
		if (n < 0) sdb_error("Can't read URL to open");
		u[n] = '\0';
		if (sdb_debuglevel) sdb_debug("'%s'\n", u);
		chomp(u);
		if (sdb_debuglevel) sdb_debug("sdb_open(%s)", u);
		db = sdb_open(u);
		if (db == NULL) {
			writedata(sock, "0", 2);
		} else {
			writedata(sock, "+0", 2);
		}
		while (doquery(sock, u) != -1);
		sdb_close(db);
	} else {
		doquery(sock, u);
	}
	if (sdb_debuglevel) sdb_debug("dourl() end");
}

static void usage(void)
{
	printf("usage: sdbd [-df] port\n");
	exit(0);
}

static void background(void)
{
	int childpid;
	if ((childpid = fork()) < 0) {
		sdb_error("Can't fork");
	} else {
		if (childpid > 0) exit(0);
	}
	setsid();
}

static int options(int argc, char **argv)
{
	int c;

	while ((c = getopt(argc, argv, "df")) != EOF) {
		switch (c) {
		case 'd':
			sdb_debuglevel++;
			break;
		case 'f':
			foreground = 1;
			break;
		default:
			usage();
		}
	}
	return optind;
}

int main(int argc, char *argv[])
{
	int sockfd, newsockfd, port, pid;
	struct sockaddr_in serv_addr, cli_addr;
	socklen_t clilen;
	int one = 1;
	int n = options(argc, argv);
	argc -= n;
	argv += n;

	if (argc < 1) {
		usage();
	}

	if (!foreground) background();

	sockfd = socket(AF_INET, SOCK_STREAM, 0);
	if (sockfd < 0) sdb_error("Error opening socket");
	memset((char *) &serv_addr, 0, sizeof(serv_addr));
	port = atoi(argv[0]);
	serv_addr.sin_family = AF_INET;
	serv_addr.sin_addr.s_addr = INADDR_ANY;
	serv_addr.sin_port = htons(port);
	setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, (char *)&one, sizeof one);
	if (bind(sockfd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
		sdb_error("Error on binding");
	listen(sockfd, 5);
	clilen = sizeof(cli_addr);
	signal(SIGCHLD, waitforchild);
	sdb_init();
	for (;;) {
		if (sdb_debuglevel) sdb_debug("accept(%d, ...)", sockfd);
		newsockfd = accept(sockfd,
				   (struct sockaddr *) &cli_addr, &clilen);
		if (newsockfd == -1) {
			if (sdb_debuglevel && errno != EINTR) {
				sdb_debug("Error on accept");
				perror("accept");
			}
			continue;
		}
		pid = fork();
		if (pid < 0) sdb_error("Error on fork");
		if (pid == 0) {
#if 1
			close(sockfd);
#endif
			dourl(newsockfd);
			exit(0);
		} else {
			close(newsockfd);
		}
	}
	return 0;
}
